pyomo/core/expr/numvalue.py

```diff
@@ -280,78 +257,74 @@ def polynomial_degree(obj):
-            try:
-                tmp = float(obj)
-                if tmp == obj:
-                    tmp = NumericConstant(tmp)
-                    update_KnownConstants(obj, tmp)
-                    return tmp
-            except:
-                pass
-
-            tmp = NumericConstant(obj)
-            update_KnownConstants(obj, tmp)
-            return tmp
-    try:
-        return obj.as_numeric()
-    except AttributeError:
-        pass
-    try:
-        if obj.__class__ is (obj + 0).__class__:
-            # obj may (or may not) be hashable, so we need this try
-            # block so that things proceed normally for non-hashable
-            # "numeric" types
-            try:
-                if obj in KnownConstants:
-                    return KnownConstants[obj]
-                else:
-                    tmp = NumericConstant(obj)
-                    update_KnownConstants(obj, tmp)
-
-                    # If we get here, this is a reasonably well-behaving
-                    # numeric type: add it to the native numeric types
-                    # so that future lookups will be faster.
-                    native_numeric_types.add(obj.__class__)
-                    # native numeric types are also native types
-                    native_types.add(obj.__class__)
-
-                    return tmp
-            except:
-                return NumericConstant(obj)
-    except:
-        pass
-    raise TypeError(
-        "Cannot convert object of type '%s' (value = %s) to a numeric value."
-        % (type(obj).__name__, obj, ))
+        #
+        # Because INT, FLOAT, and sometimes LONG hash the same, we
+        # index _KnownConstants with both the class and value.  This ensures
+        # consistent results.
+        #
+        val = _KnownConstants.get((obj.__class__, obj), None)
+        if val is not None:
+            return val 
+        #
+        # Create the numeric constant.  This really
+        # should be the only place in the code
+        # where these objects are constructed.
+        #
+        retval = NumericConstant(obj)
+        #
+        # Cache the numeric constants.  We used a bounded the cache size to avoid
+        # unexpectedly large lists of constants.  There are typically a small number
+        # of constants that need to be cached.
+        #
+        # NOTE:  A LFU policy might be more sensible here, but that
+        # requires a more complex cache.  It's not clear that that
+        # is worth the extra cost.
+        #
+        if len(_KnownConstants) < 100:
+            _KnownConstants[obj.__class__,obj] = retval
+            return retval
+        #
+        return retval
+    elif obj.__class__ in native_types:
+        raise TypeError("Cannot treat the value '%s' as a constant" % str(obj))
+    elif not isinstance_NumericValue(obj):
+        raise TypeError("Cannot treat the value '%s' as a constant because it has unknown type '%s'" % (str(obj), obj.__class__))
```

Comment:
Why isinstance?  In the old design, we did the equivalent of the following:
```
try:
    return obj.as_numeric()
except AttributeError:
    # try to figure out what to do with this type.  If it is a "new" numeric type add it to the native_numeric_types set and return it.  If it isn't, add it to the native_types set and raise an exception.
```
That is, if the object contained an `as_numeric` attribute, then we took whatever that method returned as the NumericValue.  That allowed both for duck-typing, and for non-NumericValue classes to be used in expressions as long as they implemented an as_numeric method that returned a valid NumericValue-like thing.